Save definitions climatic summary dialog

Rockjunior opened the pull request:
@lilyclements @Vitalis95
I have created a new PR to replace the previous one => #10079, as I realized there were some mistakes, misconfiguration, and possibly incorrect variable assignments in the earlier submission. I would prefer to proceed with this updated PR.

Rockjunior commented:
@Vitalis95 @N-thony 
I’m facing an issue with `ucrSaveObject `, the text is getting cut off, and the input box for **Definitions** appears too small. I tried expanding it but didn’t get it right. Could you please advise on how to fix this?

rdstern commented:
@lilyclements and @Rockjunior I had a quick look at where you have got to.
a) I note there is a checkbox called Definitions.  I wonder if it would be clearer to say `Store Definitions:  `
b) When you check the Definitions it shows the next line, see below:
<img width="670" height="845" alt="image" src="https://github.com/user-attachments/assets/209232b3-3ca5-4a5b-a3ba-2c7fad20cb03" />

You note above that there is not enough space for the words Annual_Definitions, etc.  You query above, on how to get more space, but do you really need that label at all.  I know these are Annual, from the button at the top.  I also know they are Definitions, from the checkbox just above.  And the underscore is also not needed here.  You can have spaced!

But I am questioning whether you need a label at all.  If not, then the drop down could go on the line above.  I am not sure what the drop-down is for yet?  So that's as far as I can suggest.  I got errors when I pressed Ok, but assume you know about that?

lilyclements commented on instat/dlgClimaticSummary.vb:
```diff
@@ -431,6 +503,46 @@ Public Class dlgClimaticSummary
+
+    Private Sub AddSaveDefinitionOptions()
+        ' calculated_data <- data_book$get_calculations("<resulting data frame>")
+        clsGetDataFrameFunction.AddParameter("summary_data", Chr(34) & ucrSaveObject.GetText() & Chr(34), iPosition:=0)
```
The parameter name is `data_name`, not `summary_data`

Rockjunior commented:
> Just some small quesitons, and now I'll try running your branch to see Have you also updated your `databook` package?

yes I had already, installed and updated earlier before this task using this;
`devtools::install_github("IDEMSInternational/databook")`

lilyclements commented:
Oh also, you want to read the exact same arguments into `data_book$preview_summary_names` that are read into `data_book$calculate_summary`

Rockjunior commented on instat/dlgClimaticSummary.vb:
```diff
@@ -431,6 +509,56 @@ Public Class dlgClimaticSummary
+            ucrBase.clsRsyntax.AddToBeforeCodes(clsGetDataFrameFunction, iPosition:=1)
+            ucrBase.clsRsyntax.AddToBeforeCodes(clsGetVariablesMetadataFunction, iPosition:=2)
+            ucrBase.clsRsyntax.AddToBeforeCodes(clsGetDailyDataCalculationFunction, iPosition:=3)
+            ucrBase.clsRsyntax.AddToBeforeCodes(clsGetSummaryVariablesFunction, iPosition:=4)
```
Got it,  I’ll remove this from the Before Codes.

Rockjunior commented on instat/dlgClimaticSummary.vb:
```diff
@@ -211,6 +238,40 @@ Public Class dlgClimaticSummary
+        clsGetSummaryVariablesFunction.AddParameter("factors", clsRFunctionParameter:=clsDefaultFactors, iPosition:=3)
+        clsGetSummaryVariablesFunction.SetAssignTo("summary_variables")
+        'daily_data_calculation
+        clsGetDailyDataCalculationFunction.SetRCommand(frmMain.clsRLink.strInstatDataObject & "$get_daily_data_calculation")
```
I have updated to `frmMain.clsRLink.strInstatDataObject & "$get_calculations"`. Also in [line 253](https://github.com/IDEMSInternational/R-Instat/pull/10120/changes/BASE..2c059a29b3ed312987f4f2e9dcdd7f2a294646f2#r2614583811)

lilyclements commented:
> It is currently on 2 rows in the dialog. Could the checkbox label become Store Definitions. Then the drop down has just the label Name:. It is on the same row as the checkbox in the dialog. And this user control is the same in each dialog that produces a definition. (So, at least the Summary, Start, End and Length and maybe Extremes and Spells, though that may be later?)

@rdstern do we want to have "Name:"? Usually with a ucrSave when there is a checkbox, we then don't get a new label appear.

This will be in Extremes in due course, I think.
And @Emmanuel-Afrifa is working on it on Length of Season in #10146

lilyclements commented on instat/dlgClimaticSummary.vb:
```diff
@@ -431,6 +509,51 @@ Public Class dlgClimaticSummary
+        Else
+            ' Remove all definition-related code if checkbox is unchecked
+            ucrBase.clsRsyntax.RemoveFromAfterCodes(clsGetClimaticSummariesFunction)
+            ucrBase.clsRsyntax.RemoveFromBeforeCodes(clsGetLinkedDataFrameFunction)
```
Here's your issue @Rockjunior, you have run it here to RemoveFromBeforeCodes, which is perfect. But, you need to also run it in you AddToAfterCodes

lilyclements reviewed (commented):
@Rockjunior firstly, this is great that you're testing this and reporting the issues!

1. I can only see two very minor things, which I've commented the code below which will fix your issues you're having so far

2. Then the other thing is something I commented above relating to this line

```r
summary_variables <- data_book$preview_summary_names(data_name="observations_unstacked_data", summaries=c("summary_mean"), factors=c("station_id", "s_year"))
```

This should be running the **exact same parameters and values** as your `data_book$calculate_summary`


3. You also need to update your `databook` package again. I found an error last week when debugging this. So you need to update your `databook` package.

lilyclements commented:
@Rockjunior Are you still getting 1 and 2 as errors, or is it just 3 now?

1. This is occurring because you are setting it as a variable not a string as you say. Do you know how to fix this?
2. You've outlined on how to fix it in there. Have you fixed it?
3. As I've said several times, `get_climatic_summaries_definition` is not in the data book. This means you do not run `data_book$get_climatic_summaries_definition` for this, but instead `get_climatic_summaries_definition`. Are you now doing this?

Rockjunior commented on instat/dlgClimaticSummary.vb:
```diff
@@ -431,6 +509,51 @@ Public Class dlgClimaticSummary
+            clsGetSummaryVariablesFunction.AddParameter("data_name", Chr(34) & strDataFrame & Chr(34), iPosition:=0)
+
+            ' Add to R syntax (so it appears in the final R command)
+            ucrBase.clsRsyntax.AddToAfterCodes(clsGetClimaticSummariesFunction)
```
Yes I did, am making a commit.

Rockjunior commented:
@lilyclements 
All changes have been committed, but the bug persists.

<img width="412" height="432" alt="image" src="https://github.com/user-attachments/assets/620afd1a-5a8f-4227-941a-aa6f7ac4a16f" />

Rockjunior commented:
Hello @lilyclements ,
I have implemented the changes and pushed a commit. 
This commit introduces the core logic for storing climatic summary definitions by adding a define_as_climatic workflow that creates and registers a temporary linked dataframe in the data book, consistent with the approach used in other summary dialogs. 

All related data preparation functions (_get_calculations, get_variables_metadata, and preview_summary_names_) now operate within this unified linked data context, and the variable selector and receiver controls are dynamically linked to the preview function to ensure consistent parameter passing with the main calculation flow. However, an outstanding issue remains: define_as_climatic fails when attempting to assign a climatic type to the generic variable selected via the elements receiver, as the R backend does not recognize the assigned type for this receiver. 

Could you please advise on the appropriate climatic type that should be implemented in this case? 

I am attaching screenshots of the error messages encountered while running the climatic summaries dialog for reference.
<img width="413" height="263" alt="Screenshot 2026-01-19 113347" src="https://github.com/user-attachments/assets/4dccd070-1dd7-4af6-b6ff-d520fb52a990" />

<img width="413" height="497" alt="Screenshot 2026-01-19 113353" src="https://github.com/user-attachments/assets/5804b1b2-31f8-46b1-ab6e-b6445a3a3dd2" />

Rockjunior commented:
> @Rockjunior great, I edited my initial message but I think you probably didn't see it as the error is coming from the part that I added in. My apologies for that.
> 
> This error is occurring because you're reading in the "linked_data_frame" in places you're meant to read in the data in the selector. It should be:
> 
> ```r
> calculations_data <- data_book$get_calculations(linked_data_frame)
> variables_metadata <- data_book$get_variables_metadata(" <data frame in selector> ")
> daily_data_calculation <- data_book$get_calculations(" <data frame in selector> ")
> 
> summary_variables <- data_book$preview_summary_names(data_name="< data in selector > ", ...)
> ```
> 
> Does that make sense?
> 
> Well done for fixing the other bits! This is looking really good, and very nearly there

I have made the changes, but there is one issue here with this bug
<img width="412" height="458" alt="image" src="https://github.com/user-attachments/assets/ad0601db-6f3f-4f57-9843-10b40e2876b7" />

lilyclements reviewed (changes requested):
To remove the rain parameter bits, I've just outlined below the bits to remove.

Then if you update your databook and it's working for you we might be good to go!

lilyclements commented:
@Rockjunior are you using the Ghana two stations data from the R-Instat library for this, or is it a different data set? Can you let me know what the data is / where you have found it (if it's not in the R-Instat library, can you email me a copy?)

lilyclements commented:
@Rockjunior can you try updating your databook and testing again? It should now work, after a fix in [191](https://github.com/IDEMSInternational/databook/pull/191)

lilyclements reviewed (dismissed):
@Rockjunior great work on this PR! Approving.

@Vitalis95 can you review the VB code?

Rockjunior commented:
@Vitalis95  @lilyclements I’ve made the changes requested by @Vitalis95 .
The updates should now be working as expected.
Kindly review and let me know if any further adjustments are needed.
